### PR TITLE
Move Fence to AP_Vehicle

### PR DESCRIPTION
**Pull request by IamPete1**
This moves fence to AP_Vehicle, it also fixes compilation without fence for all vehicles ~except rover were it is wrapped up in `AR_WPNav`.~

~Note that the params have not moved, so there is no need for param conversion. That could be a future improvement.~ Params are now moved and converted.

~This will add fence for tracker, will need to add dummy methods to save flash.~ Dummy methods added, I suspect it will still cost quite abit...

Should un-block https://github.com/ArduPilot/ardupilot/pull/18892 since now none-vehicle library's can tell if fence is there or not.

### CODE REVIEW

**Comment by IamPete1** [comment]
We could take the opportunity to swap from `AC_FENCE` to `AC_FENCE_ENABLED` to match the new pattern.

**Comment by IamPete1** [comment]
Elf diff of Pixhawk1-1M tracker here: https://drive.google.com/file/d/1e-HR_mVmX7r-OY8aBpDcIn6RAdX_qDKG/view?usp=sharing

After this PR We can save approx 14K by compiling out fence. We save about 6k by keeping it in with the dummy methods. 

This costs a few hundred bytes due to the param conversion on everything else. Except copter on 2M boards where is is adding 1.2K.

Durandal:
```
Binary Name      Text [B]          Data [B]       BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -------------  -------------  ----------------------------  -------------------------
antennatracker   -6536 (-0.4829%)  -4 (-0.2198%)  0 (0.0000%)    -6540 (-0.4826%)                                 617352
blimp            -6624 (-0.5152%)  -4 (-0.2203%)  0 (0.0000%)    -6628 (-0.5147%)                                 685056
ardurover        136 (+0.0085%)    0 (0.0000%)    0 (0.0000%)    136 (+0.0085%)                                   370640
arducopter-heli  1192 (+0.0666%)   0 (0.0000%)    0 (0.0000%)    1192 (+0.0665%)                                  171524
ardusub          -96 (-0.0061%)    0 (0.0000%)    0 (0.0000%)    -96 (-0.0061%)                                   392192
arduplane        72 (+0.0041%)     0 (0.0000%)    0 (0.0000%)    72 (+0.0041%)                                    204104
arducopter       1236 (+0.0695%)   0 (0.0000%)    -4 (-0.0015%)  1236 (+0.0694%)                                  182644
```

MatekF405:
```
Binary Name      Text [B]          Data [B]       BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ----------------  -------------  -------------  ----------------------------  -------------------------
antennatracker   -5816 (-0.8796%)  -4 (-0.5236%)  0 (0.0000%)    -5820 (-0.8792%)                                 326904
blimp            -5860 (-0.8622%)  -4 (-0.5263%)  -4 (-0.0031%)  -5864 (-0.8618%)                                 308456
ardurover        16 (+0.0019%)     0 (0.0000%)    0 (0.0000%)    16 (+0.0019%)                                    142816
arducopter-heli  196 (+0.0209%)    0 (0.0000%)    4 (+0.0031%)   196 (+0.0208%)                                    40956
ardusub          112 (+0.0138%)    0 (0.0000%)    0 (0.0000%)    112 (+0.0138%)                                   171984
arduplane        148 (+0.0154%)    0 (0.0000%)    4 (+0.0031%)   148 (+0.0153%)                                    16784
arducopter       184 (+0.0199%)    0 (0.0000%)    0 (0.0000%)    184 (+0.0198%)                                    53980
```

File: Tools/ardupilotwaf/ardupilotwaf.py [hunk 1]
```diff
@@ -109,7 +109,8 @@
     'AP_VideoTX',
     'AP_FETtecOneWire',
     'AP_Torqeedo',
-    'AP_CustomRotations'
+    'AP_CustomRotations',
+    'AC_Fence'
```
Comment (IamPete1) [inline Tools/ardupilotwaf/ardupilotwaf.py:113]: don't need this, its already on 88

**Comment by tridge** [comment]
need to work out why fw size changes are so strange

**Comment by IamPete1** [comment]
A rebase seems to have fixed the random increase in flash we were seeing on copter.

**Review by khancyr — APPROVED** [review]
Looks good.

File: libraries/AC_Fence/AC_Fence.h [hunk 2]
```diff
@@ -35,6 +39,8 @@
 #define AC_FENCE_GIVE_UP_DISTANCE                   100.0f  // distance outside the fence at which we should give up and just land.  Note: this is not used by library directly but is intended to be used by the main code
 #define AC_FENCE_MANUAL_RECOVERY_TIME_MIN           10000   // pilot has 10seconds to recover during which time the autopilot will not attempt to re-take control
 
+#if AC_FENCE
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/AC_Fence/AC_Fence.h]: done, thanks

File: ArduCopter/Copter.h [hunk 3]
```diff
@@ -119,9 +119,7 @@
 #if MODE_FOLLOW_ENABLED == ENABLED
  # include <AP_Follow/AP_Follow.h>
 #endif
-#if AC_FENCE == ENABLED
  # include <AC_Fence/AC_Fence.h>
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply ArduCopter/Copter.h]: yep, done.

**Comment by IamPete1** [comment]
Rebased, moved to `AC_Fence_config.h`  and renamed `AC_FENCE` to `AP_FENCE_ENABLED`

**Comment by tridge** [comment]
it would be good to add to build_options.py, so we know that building without fence keeps working

**Comment by tridge** [comment]
just need some extensive testing in SITL with various fence options

**Comment by tridge** [comment]
i think we can get this in for 4.3.x, as long as it has enough testing

**Comment by IamPete1** [comment]
Manually tested fence on Copter, Plane, Rover and sub. Failsafes trigger as expected.

File: ArduCopter/Parameters.h [hunk 4]
```diff
@@ -143,7 +143,7 @@ class Parameters {
         k_param_gpslock_limit,          // deprecated - remove
         k_param_geofence_limit,         // deprecated - remove
         k_param_altitude_limit,         // deprecated - remove
-        k_param_fence,
+        k_param_fence,                  // deprecated - remove
```
Comment (peterbarker) [inline ArduCopter/Parameters.h:146]:
```suggestion
        k_param_fence_old,                  // only used for conversion
```

lets the compiler help you on any usage - and we can't reuse/remove :-)

File: libraries/AC_Fence/AC_Fence_config.h [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#pragma once
+
+#include <AP_HAL/AP_HAL_Boards.h>
+
+#ifndef AP_FENCE_ENABLED
+#define AP_FENCE_ENABLED 2
```
Comment (tridge) [inline libraries/AC_Fence/AC_Fence_config.h:6]: comment explaining what 2 means

File: Tools/scripts/build_options.py [hunk 6]
```diff
@@ -35,8 +35,10 @@ def __init__(self,
     Feature('AHRS', 'VISUALODOM', 'HAL_VISUALODOM_ENABLED', 'Enable Visual Odometry', 0, None),
 
     Feature('Safety', 'PARACHUTE', 'HAL_PARACHUTE_ENABLED', 'Enable Parachute', 0, None),
+    Feature('Safety', 'FENCE', 'AP_FENCE_ENABLED', 'Enable Geofence', 0, None),
```
Comment (tridge) [inline Tools/scripts/build_options.py:38]: default fence to on

File: ArduSub/Parameters.h [hunk 7]
```diff
@@ -87,7 +87,7 @@ class Parameters {
         k_param_pos_control, // Position Control
         k_param_wp_nav, // Waypoint navigation
         k_param_mission, // Mission library
-        k_param_fence, // Fence Library
+        k_param_fence, // unused
```
Comment (peterbarker) [inline ArduSub/Parameters.h:90]:
```suggestion
        k_param_fence, // used only for conversion
```

File: libraries/AC_Avoidance/AC_Avoid.cpp [hunk 8]
```diff
@@ -152,6 +153,7 @@ void AC_Avoid::adjust_velocity_fence(float kP, float accel_cmss, Vector3f &desir
         adjust_velocity_exclusion_circles(kP, accel_cmss_limited, desired_velocity_xy_cms, backup_vel_fence, dt);
         find_max_quadrant_velocity(backup_vel_fence, quad_1_back_vel, quad_2_back_vel, quad_3_back_vel, quad_4_back_vel);
     }
+#endif // AC_FENCE
```
Comment (peterbarker) [inline libraries/AC_Avoidance/AC_Avoid.cpp:156]:
```suggestion
#endif // AP_FENCE_ENABLED
```

File: ArduCopter/Copter.h [hunk 9]
```diff
@@ -160,6 +157,14 @@
 #include <AP_Scripting/AP_Scripting.h>
 #endif
 
+#if AC_AVOID_ENABLED && !AP_FENCE_ENABLED
+  #warning AC_Avoidance relies on AP_FENCE_ENABLED which is disabled
```
Comment (tridge) [inline ArduCopter/Copter.h:161]: better to be an #error

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Review by tridge — CHANGES_REQUESTED** [review]

File: Tools/scripts/build_options.py (hunk 6 above)
Comment (IamPete1) [reply Tools/scripts/build_options.py]: Defaulted to value of 2 to get vehicle dependent dummy methods, the same as the hard-coded default. I'm not sure if the server can handle a value of 2?

**Review by tridge — APPROVED** [review]

**Pull request merged**